source/vof/handler.cc

```diff
@@ -0,0 +1,358 @@
+
+      if (x_vof_composition_vars.size()>0 && !sim.parameters.use_discontinuous_composition_discretization)
+        {
+          AssertThrow(false, ExcMessage("Volume of Fluid composition field not implemented for continuous composition."));
+        }
+
+      for (std::vector<std::string>::const_iterator p = x_vof_composition_vars.begin();
+           p != x_vof_composition_vars.end(); ++p)
+        {
+          const std::vector<std::string> split_parts = Utilities::split_string_list(*p, ':');
+          AssertThrow (split_parts.size() == 2,
+                       ExcMessage("The format for Volume of Fluid composition mappings requires that each entry has the form"
+                                  "`composition:Vof field', but there does not appear to be a colon in the entry <" + *p + ">."));
+
+          const std::string composition_field = split_parts[0];
+          const std::string vof_field = split_parts[1];
+
+          // Check composition_field exists
+          bool field_exists=false;
+
+          for (unsigned int i=0; i<sim.parameters.n_compositional_fields; ++i)
+            {
+              field_exists = field_exists ||
+                             (composition_field==sim.parameters.names_of_compositional_fields[i]);
+            }
+
+          Assert(field_exists, ExcMessage("Composition field variable " +
+                                          composition_field +
+                                          " does not exist."));
+
+          // Check vof_field exists
+          field_exists=false;
+          unsigned int vof_field_index = n_vof_fields;
+
+          for (unsigned int i=0; i<n_vof_fields; ++i)
+            {
+              if (vof_field == vof_field_names[i])
+                {
+                  field_exists = true;
+                  vof_field_index = i;
+                  break;
+                }
+            }
+
+          Assert(field_exists, ExcMessage("Volume of Fluid field variable " +
+                                          vof_field +
+                                          " does not exist."));
+
+          // Ensure no duplicate mapping to a composition field
+          Assert (vof_composition_map_index.count(composition_field) == 0,
+                  ExcMessage("Volume of Fluid composition field mappings have to be unique! " + composition_field +
+                             " is used more than once."));
+
+
+          // Add to mappings
+          vof_composition_map_index[composition_field] = vof_field_index;
+        }
+    }
+    prm.leave_subsection ();
+  }
+
+  template <int dim>
+  void
+  VoFHandler<dim>::initialize (ParameterHandler &prm)
+  {
+    // Do checks on required assumptions
+    AssertThrow(dim==2,ExcMessage("Volume of Fluid Interface Tracking is currently only functional for dim=2."));
+    AssertThrow(this->get_parameters().CFL_number < 1.0, ExcMessage("Volume of Fluid Interface Tracking requires CFL < 1."));
+
```

Comment:
Add an `AssertThrow(dynamic_cast<MappingCartesian<dim> *> (&this->get_mapping()) != 0, ExcMessage(...))`